File: android/app/src/main/java/com/phoneintegration/app/models/SyncFlowCall.kt
package com.phoneintegration.app.models

/**
 * Represents a SyncFlow-to-SyncFlow audio/video call between devices.
 * This is separate from cellular phone calls - it's app-to-app VoIP.
 *
 * VPS Backend Only - Firebase annotations removed.
 */
data class SyncFlowCall(
    val id: String = "",
    val callerId: String = "",
    val callerName: String = "",
    val callerPlatform: String = "",
    val calleeId: String = "",
    val calleeName: String = "",
    val calleePlatform: String = "",
    val callType: CallType = CallType.AUDIO,
    val status: CallStatus = CallStatus.RINGING,
    val startedAt: Long = 0,
    val answeredAt: Long? = null,
    val endedAt: Long? = null,
    val offer: SDPData? = null,
    val answer: SDPData? = null,
    // New fields for user-to-user calling
    val isUserCall: Boolean = false,
    val callerPhone: String = ""
) {
    enum class CallType {
        AUDIO,
        VIDEO;

        companion object {
            fun fromString(value: String): CallType {
                return when (value.lowercase()) {
                    "video" -> VIDEO
                    else -> AUDIO
                }
            }
        }

        override fun toString(): String = name.lowercase()
    }

    enum class CallStatus {
        RINGING,    // Call initiated, waiting for answer
        ACTIVE,     // Call in progress
        ENDED,      // Call ended normally
        REJECTED,   // Callee rejected the call
        MISSED,     // Call timed out without answer
        FAILED;     // Call failed to connect

        companion object {
            fun fromString(value: String): CallStatus {
                return when (value.lowercase()) {
                    "ringing" -> RINGING
                    "active" -> ACTIVE
                    "ended" -> ENDED
                    "rejected" -> REJECTED
                    "missed" -> MISSED
                    "failed" -> FAILED
                    else -> RINGING
                }
            }
        }

        override fun toString(): String = name.lowercase()
    }

    /**
     * SDP (Session Description Protocol) data for WebRTC signaling
     */
    data class SDPData(
        val sdp: String = "",
        val type: String = ""
    )

    /**
     * ICE (Interactive Connectivity Establishment) candidate for WebRTC
     */
    data class IceCandidate(
        val candidate: String = "",
        val sdpMid: String? = null,
        val sdpMLineIndex: Int = 0
    )

    /**
     * Convert to map for API serialization
     */
    fun toMap(): Map<String, Any?> = mapOf(
        "id" to id,
        "callerId" to callerId,
        "callerName" to callerName,
        "callerPlatform" to callerPlatform,
        "calleeId" to calleeId,
        "calleeName" to calleeName,
        "calleePlatform" to calleePlatform,
        "callType" to callType.toString(),
        "status" to status.toString(),
        "startedAt" to startedAt,
        "answeredAt" to answeredAt,
        "endedAt" to endedAt,
        "offer" to offer?.let { mapOf("sdp" to it.sdp, "type" to it.type) },
        "answer" to answer?.let { mapOf("sdp" to it.sdp, "type" to it.type) },
        "isUserCall" to isUserCall,
        "callerPhone" to callerPhone
    )

    val isIncoming: Boolean
        get() = callerPlatform != "android"

    val isOutgoing: Boolean
        get() = callerPlatform == "android"

    val isVideo: Boolean
        get() = callType == CallType.VIDEO

    val isActive: Boolean
        get() = status == CallStatus.ACTIVE

    val isRinging: Boolean
        get() = status == CallStatus.RINGING

    val displayName: String
        get() = if (isIncoming) callerName else calleeName

    val duration: Long
        get() {
            val start = answeredAt ?: return 0
            val end = endedAt ?: System.currentTimeMillis()
            return end - start
        }

    companion object {
        const val CALL_TIMEOUT_MS = 60_000L // 60 seconds

        /**
         * Create a new outgoing call from this device
         */
        fun createOutgoing(
            callId: String,
            callerId: String,
            callerName: String,
            calleeId: String,
            calleeName: String,
            isVideo: Boolean
        ): SyncFlowCall = SyncFlowCall(
            id = callId,
            callerId = callerId,
            callerName = callerName,
            callerPlatform = "android",
            calleeId = calleeId,
            calleeName = calleeName,
            calleePlatform = "macos",
            callType = if (isVideo) CallType.VIDEO else CallType.AUDIO,
            status = CallStatus.RINGING,
            startedAt = System.currentTimeMillis()
        )

        /**
         * Parse from map data
         */
        fun fromMap(id: String, map: Map<String, Any?>): SyncFlowCall {
            val offerMap = map["offer"] as? Map<*, *>
            val answerMap = map["answer"] as? Map<*, *>

            return SyncFlowCall(
                id = id,
                callerId = map["callerId"] as? String ?: map["callerUid"] as? String ?: "",
                callerName = map["callerName"] as? String ?: "",
                callerPlatform = map["callerPlatform"] as? String ?: "",
                calleeId = map["calleeId"] as? String ?: "",
                calleeName = map["calleeName"] as? String ?: "",
                calleePlatform = map["calleePlatform"] as? String ?: "",
                callType = CallType.fromString(map["callType"] as? String ?: "audio"),
                status = CallStatus.fromString(map["status"] as? String ?: "ringing"),
                startedAt = (map["startedAt"] as? Number)?.toLong() ?: 0,
                answeredAt = (map["answeredAt"] as? Number)?.toLong(),
                endedAt = (map["endedAt"] as? Number)?.toLong(),
                offer = offerMap?.let {
                    SDPData(
                        sdp = it["sdp"] as? String ?: "",
                        type = it["type"] as? String ?: ""
                    )
                },
                answer = answerMap?.let {
                    SDPData(
                        sdp = it["sdp"] as? String ?: "",
                        type = it["type"] as? String ?: ""
                    )
                },
                isUserCall = map["isUserCall"] as? Boolean ?: false,
                callerPhone = map["callerPhone"] as? String ?: ""
            )
        }
    }
}

/**
 * Device information for SyncFlow calling
 */
data class SyncFlowDevice(
    val id: String = "",
    val name: String = "",
    val platform: String = "",
    val online: Boolean = false,
    val lastSeen: Long = 0
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "id" to id,
        "name" to name,
        "platform" to platform,
        "online" to online,
        "lastSeen" to lastSeen
    )

    val isMacOS: Boolean
        get() = platform == "macos"

    val isAndroid: Boolean
        get() = platform == "android"

    companion object {
        fun fromMap(id: String, map: Map<String, Any?>): SyncFlowDevice = SyncFlowDevice(
            id = id,
            name = map["name"] as? String ?: "",
            platform = map["platform"] as? String ?: "",
            online = map["online"] as? Boolean ?: false,
            lastSeen = (map["lastSeen"] as? Number)?.toLong() ?: 0
        )
    }
}

File: SyncFlowMac/SyncFlowMac/Models/SyncFlowCall.swift
//
//  SyncFlowCall.swift
//  SyncFlowMac
//
//  Model for SyncFlow-to-SyncFlow audio/video calls between devices.
//  This is separate from cellular phone calls - it's app-to-app VoIP.
//

import Foundation

/// Represents a SyncFlow audio/video call between devices
struct SyncFlowCall: Identifiable, Equatable {
    let id: String
    let callerId: String
    let callerName: String
    let callerPlatform: String
    let calleeId: String
    let calleeName: String
    let calleePlatform: String
    let callType: CallType
    var status: CallStatus
    let startedAt: Date
    var answeredAt: Date?
    var endedAt: Date?
    var offer: SDPData?
    var answer: SDPData?

    enum CallType: String, Codable {
        case audio
        case video

        var isVideo: Bool { self == .video }
    }

    enum CallStatus: String, Codable {
        case ringing
        case active
        case ended
        case rejected
        case missed
        case failed
    }

    /// SDP (Session Description Protocol) data for WebRTC signaling
    struct SDPData: Equatable {
        let sdp: String
        let type: String
    }

    /// ICE (Interactive Connectivity Establishment) candidate for WebRTC
    struct IceCandidate: Equatable {
        let candidate: String
        let sdpMid: String?
        let sdpMLineIndex: Int

        func toDict() -> [String: Any] {
            return [
                "candidate": candidate,
                "sdpMid": sdpMid ?? "",
                "sdpMLineIndex": sdpMLineIndex
            ]
        }

        static func from(_ dict: [String: Any]) -> IceCandidate? {
            guard let candidate = dict["candidate"] as? String else { return nil }
            return IceCandidate(
                candidate: candidate,
                sdpMid: dict["sdpMid"] as? String,
                sdpMLineIndex: (dict["sdpMLineIndex"] as? Int) ?? 0
            )
        }
    }

    // Computed properties
    var isIncoming: Bool { callerPlatform != "macos" }
    var isOutgoing: Bool { callerPlatform == "macos" }
    var isVideoCall: Bool { callType == .video }
    var isActive: Bool { status == .active }
    var isRinging: Bool { status == .ringing }

    var displayName: String {
        isIncoming ? callerName : calleeName
    }

    var duration: TimeInterval {
        guard let start = answeredAt else { return 0 }
        let end = endedAt ?? Date()
        return end.timeIntervalSince(start)
    }

    var formattedDuration: String {
        let totalSeconds = Int(duration)
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }

    // MARK: - Factory Methods

    /// Create an outgoing call from this Mac
    static func createOutgoing(
        callId: String,
        callerId: String,
        callerName: String,
        calleeId: String,
        calleeName: String,
        isVideo: Bool
    ) -> SyncFlowCall {
        return SyncFlowCall(
            id: callId,
            callerId: callerId,
            callerName: callerName,
            callerPlatform: "macos",
            calleeId: calleeId,
            calleeName: calleeName,
            calleePlatform: "android",
            callType: isVideo ? .video : .audio,
            status: .ringing,
            startedAt: Date(),
            answeredAt: nil,
            endedAt: nil,
            offer: nil,
            answer: nil
        )
    }

    /// Parse from dictionary
    static func from(id: String, dict: [String: Any]) -> SyncFlowCall? {
        guard let callerId = dict["callerId"] as? String,
              let callerName = dict["callerName"] as? String,
              let callerPlatform = dict["callerPlatform"] as? String,
              let calleeId = dict["calleeId"] as? String,
              let calleeName = dict["calleeName"] as? String,
              let calleePlatform = dict["calleePlatform"] as? String,
              let callTypeStr = dict["callType"] as? String,
              let statusStr = dict["status"] as? String,
              let startedAtMs = dict["startedAt"] as? Double
        else {
            return nil
        }

        let callType = CallType(rawValue: callTypeStr) ?? .audio
        let status = CallStatus(rawValue: statusStr) ?? .ringing
        let startedAt = Date(timeIntervalSince1970: startedAtMs / 1000)

        var answeredAt: Date?
        if let answeredAtMs = dict["answeredAt"] as? Double {
            answeredAt = Date(timeIntervalSince1970: answeredAtMs / 1000)
        }

        var endedAt: Date?
        if let endedAtMs = dict["endedAt"] as? Double {
            endedAt = Date(timeIntervalSince1970: endedAtMs / 1000)
        }

        var offer: SDPData?
        if let offerDict = dict["offer"] as? [String: Any],
           let sdp = offerDict["sdp"] as? String,
           let type = offerDict["type"] as? String {
            offer = SDPData(sdp: sdp, type: type)
        }

        var answer: SDPData?
        if let answerDict = dict["answer"] as? [String: Any],
           let sdp = answerDict["sdp"] as? String,
           let type = answerDict["type"] as? String {
            answer = SDPData(sdp: sdp, type: type)
        }

        return SyncFlowCall(
            id: id,
            callerId: callerId,
            callerName: callerName,
            callerPlatform: callerPlatform,
            calleeId: calleeId,
            calleeName: calleeName,
            calleePlatform: calleePlatform,
            callType: callType,
            status: status,
            startedAt: startedAt,
            answeredAt: answeredAt,
            endedAt: endedAt,
            offer: offer,
            answer: answer
        )
    }

    /// Convert to dictionary
    func toDict() -> [String: Any] {
        var dict: [String: Any] = [
            "id": id,
            "callerId": callerId,
            "callerName": callerName,
            "callerPlatform": callerPlatform,
            "calleeId": calleeId,
            "calleeName": calleeName,
            "calleePlatform": calleePlatform,
            "callType": callType.rawValue,
            "status": status.rawValue,
            "startedAt": startedAt.timeIntervalSince1970 * 1000
        ]

        if let answeredAt = answeredAt {
            dict["answeredAt"] = answeredAt.timeIntervalSince1970 * 1000
        }
        if let endedAt = endedAt {
            dict["endedAt"] = endedAt.timeIntervalSince1970 * 1000
        }
        if let offer = offer {
            dict["offer"] = ["sdp": offer.sdp, "type": offer.type]
        }
        if let answer = answer {
            dict["answer"] = ["sdp": answer.sdp, "type": answer.type]
        }

        return dict
    }

    // MARK: - Constants

    static let callTimeoutSeconds: TimeInterval = 60
}

/// Device information for SyncFlow calling
struct SyncFlowDevice: Identifiable, Equatable {
    let id: String
    let name: String
    let platform: String
    let online: Bool
    let lastSeen: Date

    var isMacOS: Bool { platform == "macos" }
    var isAndroid: Bool { platform == "android" }

    static func from(id: String, dict: [String: Any]) -> SyncFlowDevice? {
        guard let name = dict["name"] as? String,
              let platform = dict["platform"] as? String
        else {
            return nil
        }

        let online = dict["online"] as? Bool ?? false
        let lastSeenMs = dict["lastSeen"] as? Double ?? 0
        let lastSeen = Date(timeIntervalSince1970: lastSeenMs / 1000)

        return SyncFlowDevice(
            id: id,
            name: name,
            platform: platform,
            online: online,
            lastSeen: lastSeen
        )
    }

    func toDict() -> [String: Any] {
        return [
            "id": id,
            "name": name,
            "platform": platform,
            "online": online,
            "lastSeen": lastSeen.timeIntervalSince1970 * 1000
        ]
    }
}
